<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Given a Pharmacoset of the sensitivity experiment type, and a list of drugs,
the function will compute a signature for the effect gene expression on the
molecular profile of a cell. The function returns the estimated coefficient,
the t-stat, the p-value and the false discovery rate associated with that
coefficient, in a 3 dimensional array, with genes in the first direction,
drugs in the second, and the selected return values in the third."><title>Creates a signature representing the association between gene expression (or
other molecular profile) and drug dose response, for use in drug sensitivity
analysis. — drugSensitivitySig,PharmacoSet-method • PharmacoGx</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Creates a signature representing the association between gene expression (or
other molecular profile) and drug dose response, for use in drug sensitivity
analysis. — drugSensitivitySig,PharmacoSet-method"><meta property="og:description" content="Given a Pharmacoset of the sensitivity experiment type, and a list of drugs,
the function will compute a signature for the effect gene expression on the
molecular profile of a cell. The function returns the estimated coefficient,
the t-stat, the p-value and the false discovery rate associated with that
coefficient, in a 3 dimensional array, with genes in the first direction,
drugs in the second, and the selected return values in the third."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">PharmacoGx</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.3.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/PharmacoGx.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/CreatingPharmacoSet.html">PharmacoGx: An R Package for Analysis of Large Pharmacogenomic Datasets</a>
    <a class="dropdown-item" href="../articles/DetectingDrugSynergyAndAntagonism.html">Detecting Drug Synergy and Antagonism with PharmacoGx 3.0+</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bhklab/PharmacoGx/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Creates a signature representing the association between gene expression (or
other molecular profile) and drug dose response, for use in drug sensitivity
analysis.</h1>
      <small class="dont-index">Source: <a href="https://github.com/bhklab/PharmacoGx/blob/HEAD/R/methods-drugSensitivitySig.R" class="external-link"><code>R/methods-drugSensitivitySig.R</code></a></small>
      <div class="d-none name"><code>drugSensitivitySig-PharmacoSet-method.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Given a Pharmacoset of the sensitivity experiment type, and a list of drugs,
the function will compute a signature for the effect gene expression on the
molecular profile of a cell. The function returns the estimated coefficient,
the t-stat, the p-value and the false discovery rate associated with that
coefficient, in a 3 dimensional array, with genes in the first direction,
drugs in the second, and the selected return values in the third.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for PharmacoSet</span></span>
<span><span class="fu">drugSensitivitySig</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">mDataType</span>,</span>
<span>  <span class="va">drugs</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  <span class="va">cells</span>,</span>
<span>  <span class="va">tissues</span>,</span>
<span>  sensitivity.measure <span class="op">=</span> <span class="st">"auc_recomputed"</span>,</span>
<span>  molecular.summary.stat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mean"</span>, <span class="st">"median"</span>, <span class="st">"first"</span>, <span class="st">"last"</span>, <span class="st">"or"</span>, <span class="st">"and"</span><span class="op">)</span>,</span>
<span>  sensitivity.summary.stat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mean"</span>, <span class="st">"median"</span>, <span class="st">"first"</span>, <span class="st">"last"</span><span class="op">)</span>,</span>
<span>  returnValues <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"estimate"</span>, <span class="st">"pvalue"</span>, <span class="st">"fdr"</span><span class="op">)</span>,</span>
<span>  <span class="va">sensitivity.cutoff</span>,</span>
<span>  standardize <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SD"</span>, <span class="st">"rescale"</span>, <span class="st">"none"</span><span class="op">)</span>,</span>
<span>  molecular.cutoff <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  molecular.cutoff.direction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"less"</span>, <span class="st">"greater"</span><span class="op">)</span>,</span>
<span>  nthread <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  parallel.on <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"drug"</span>, <span class="st">"gene"</span><span class="op">)</span>,</span>
<span>  modeling.method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"anova"</span>, <span class="st">"pearson"</span><span class="op">)</span>,</span>
<span>  inference.method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"analytic"</span>, <span class="st">"resampling"</span><span class="op">)</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>object</dt>
<dd><p><code>PharmacoSet</code> a PharmacoSet of the perturbation experiment type</p></dd>


<dt>mDataType</dt>
<dd><p><code>character</code> which one of the molecular data types to use
in the analysis, out of dna, rna, rnaseq, snp, cnv</p></dd>


<dt>drugs</dt>
<dd><p><code>character</code> a vector of drug names for which to compute the
signatures. Should match the names used in the PharmacoSet.</p></dd>


<dt>features</dt>
<dd><p><code>character</code> a vector of features for which to compute the
signatures. Should match the names used in correspondant molecular data in PharmacoSet.</p></dd>


<dt>cells</dt>
<dd><p><code>character</code> allows choosing exactly which cell lines to include for the signature fitting.
Should be a subset of sampleNames(pSet)</p></dd>


<dt>tissues</dt>
<dd><p><code>character</code> a vector of which tissue types to include in the signature fitting.
Should be a subset of sampleInfo(pSet)$tissueid</p></dd>


<dt>sensitivity.measure</dt>
<dd><p><code>character</code> which measure of the drug dose
sensitivity should the function use for its computations? Use the
sensitivityMeasures function to find out what measures are available for each PSet.</p></dd>


<dt>molecular.summary.stat</dt>
<dd><p><code>character</code> What summary statistic should be used to
summarize duplicates for cell line molecular profile measurements?</p></dd>


<dt>sensitivity.summary.stat</dt>
<dd><p><code>character</code> What summary statistic should be used to
summarize duplicates for cell line sensitivity measurements?</p></dd>


<dt>returnValues</dt>
<dd><p><code>character</code> Which of estimate, t-stat, p-value and fdr
should the function return for each gene drug pair?</p></dd>


<dt>sensitivity.cutoff</dt>
<dd><p><code>numeric</code> Allows the user to binarize the sensitivity data using this threshold.</p></dd>


<dt>standardize</dt>
<dd><p><code>character</code> One of "SD", "rescale", or "none", for the form of standardization of
the data to use. If "SD", the the data is scaled so that SD = 1. If rescale, then the data is scaled so that the 95%
interquantile range lies in [0,1]. If none no rescaling is done.</p></dd>


<dt>molecular.cutoff</dt>
<dd><p>Allows the user to binarize the sensitivity data using this threshold.</p></dd>


<dt>molecular.cutoff.direction</dt>
<dd><p><code>character</code> One of "less" or "greater", allows to set direction of binarization.</p></dd>


<dt>nthread</dt>
<dd><p><code>numeric</code> if multiple cores are available, how many cores
should the computation be parallelized over?</p></dd>


<dt>parallel.on</dt>
<dd><p>One of "gene" or "drug", chooses which level to parallelize computation (by gene, or by drug).</p></dd>


<dt>modeling.method</dt>
<dd><p>One of "anova" or "pearson". If "anova", nested linear models (including and excluding the molecular feature) adjusted for
are fit after the data is standardized, and ANOVA is used to estimate significance. If "pearson", partial correlation adjusted for tissue of origin are
fit to the data, and a Pearson t-test (or permutation) test are used. Note that the difference is in whether standardization is done across the whole
dataset (anova) or within each tissue (pearson), as well as the test applied.</p></dd>


<dt>inference.method</dt>
<dd><p>Should "analytic" or "resampling" (permutation testing + bootstrap) inference be used to estimate significance.
For permutation testing, QUICK-STOP is used to adaptively stop permutations. Resampling is currently only implemented for "pearson" modelling method.</p></dd>


<dt>verbose</dt>
<dd><p><code>logical</code> 'TRUE' if the warnings and other informative message shoud be displayed</p></dd>


<dt>...</dt>
<dd><p>additional arguments not currently fully supported by the function</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p><code>array</code> a 3D array with genes in the first dimension, drugs in the
second, and return values in the third.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">GDSCsmall</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">drug.sensitivity</span> <span class="op">&lt;-</span> <span class="fu">drugSensitivitySig</span><span class="op">(</span><span class="va">GDSCsmall</span>,</span></span>
<span class="r-in"><span>  mDataType <span class="op">=</span> <span class="st">"rna"</span>,</span></span>
<span class="r-in"><span>  nthread <span class="op">=</span> <span class="fl">1</span>, features <span class="op">=</span> <span class="fu"><a href="PharmacoSet-accessors.html">fNames</a></span><span class="op">(</span><span class="va">GDSCsmall</span>, <span class="st">"rna"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Summarizing rna molecular data for:	GDSC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                                              |                                                                      |   0%  |                                                                              |======================================================================| 100%</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Computing drug sensitivity signatures...</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">drug.sensitivity</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> PharmacoSet Name:  GDSC </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Signature Type:  Sensitivity </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Date Created:  Mon Dec 12 18:21:08 2022 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of Drugs:  139 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of Genes/Probes:  1 </span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Petr Smirnov, Christopher Eeles, Zhaleh Safikhani, Mark Freeman, Feifei Li, Benjamin Haibe-Kains.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer></div>

  

  

  </body></html>

