<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="PharmacoGx">
<title>PharmacoGx: An R Package for Analysis of Large Pharmacogenomic Datasets • PharmacoGx</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="PharmacoGx: An R Package for Analysis of Large Pharmacogenomic Datasets">
<meta property="og:description" content="PharmacoGx">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">PharmacoGx</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.1.7</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/PharmacoGx.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/CreatingPharmacoSet.html">PharmacoGx: An R Package for Analysis of Large Pharmacogenomic Datasets</a>
    <a class="dropdown-item" href="../articles/DetectingDrugSynergyAndAntagonism.html">Detecting Drug Synergy and Antagonism with PharmacoGx 3.0+</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bhklab/PharmacoGx/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>PharmacoGx: An R Package for Analysis of Large Pharmacogenomic Datasets</h1>
                        <h4 data-toc-skip class="author">Petr
Smirnov</h4>
            <address class="author_afil">
      Princess Margaret Cancer Centre, University Health Network,
Toronto Canada<br><a class="author_email" href="mailto:#"></a><a href="mailto:petr.smirnov@uhnresearch.ca" class="email">petr.smirnov@uhnresearch.ca</a>
      </address>
                              <h4 data-toc-skip class="author">Christopher
Eeles</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:christopher.eeles@uhnresearch.ca" class="email">christopher.eeles@uhnresearch.ca</a>
      
                              <h4 data-toc-skip class="author">Zhaleh
Safikhani</h4>
            <address class="author_afil">
      Princess Margaret Cancer Centre, University Health Network,
Toronto CanadaDepartment of Medical Biophysics, University of Toronto,
Toronto
Canada<br><h4 data-toc-skip class="author">Benjamin
Haibe-Kains</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:benjamin.haibe.kains@utoronto.ca" class="email">benjamin.haibe.kains@utoronto.ca</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/bhklab/PharmacoGx/blob/HEAD/vignettes/CreatingPharmacoSet.Rmd" class="external-link"><code>vignettes/CreatingPharmacoSet.Rmd</code></a></small>
      <div class="d-none name"><code>CreatingPharmacoSet.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The PharmacoSet class is structured to contain several different
types of genomic data, as well as drug dose response data. It is built
upon base R data types and the <code>SummarizedExperiment</code> class.
However, a <code>PharmacoSet</code> object requires the presence of
specific annotations for the data to be able to function correctly with
the function provided in PharmacoGx. The structure allows users to then
interrogate the data on the basis of cells and drugs, removing the need
for the user to deal with single experiments. We do not recommend
creating PharmacoSet objects by the end users, rather we recommend
contacting us to have the objects created for your dataset of interest.
However, for completeness and power users the process and structure is
documented below.</p>
</div>
<div class="section level2">
<h2 id="pharmacoset-structure">PharmacoSet Structure<a class="anchor" aria-label="anchor" href="#pharmacoset-structure"></a>
</h2>
<p>The base PharmacoSet structure is as follows:</p>
<div class="section level3">
<h3 id="sample">
<code>sample</code><a class="anchor" aria-label="anchor" href="#sample"></a>
</h3>
<p>Previously called <code>@cell</code>.</p>
<p>This slot contains a <code>data.frame</code> which contains
information about all the cells profiles across all the data types in
the PharamcoSet, including both perturbation and sensitivity
experiments. It is crucial for the rownames of each entry in the data
frame to be the <strong>unique cell identifier</strong> used across all
the datatypes for each cell type. The content of this data frame will
vary based on what information each dataset provides. Mandatory columns
in this data frame must be <code>sampleid</code>, which should be
equivalent to the rownames and <code>tissueid</code>, which contains a
curated annotation of the source tissue for each cell line.</p>
</div>
<div class="section level3">
<h3 id="treatment">
<code>treatment</code><a class="anchor" aria-label="anchor" href="#treatment"></a>
</h3>
<p>Previously called <code>@drug</code>.</p>
<p>This slot contains a <code>data.frame</code> which contains
information about all the drugs profiled across all the data types in
the PharmacoSet, including both perturbation and sensitivity
experiments. Similar to the <code>sample</code> slot, it is crucial for
the rownames of each entry in the data frame to be the <strong>unique
compound identifier</strong> used across all the datatypes for each
compound. Once again, the content in this data frame will vary based on
information provided by each dataset. Mandatory columns must be
<code>treatmentid</code>, which should be equivalent to the object
rownames.</p>
</div>
<div class="section level3">
<h3 id="curation">
<code>curation</code>:<a class="anchor" aria-label="anchor" href="#curation"></a>
</h3>
<p>This slot contains three data.frames, one for drugs, tissues, and
cells. Each contains two columns, the first with the unique identifier
that is used between all PharmacoSet objects for this drug, and the
second with the identifier used within the dataset. At this time there
is no set method to find the unique identifiers between all PharmacoSets
for arbitrary drugs, as much labour intensive curation has gone into
carefully matching identifiers between datasets. However, these tables
are used only by the intersecting function, and all other functions will
function without this data, therefore for PharmacoSet objects created by
users, we recommend just filling both columns with the identifiers used
in the datasets, to allow for at least some matching to be done.</p>
<ul>
<li>
<code>sample</code>: data frame annotating all cell-lines
investigated in the study.</li>
<li>
<code>treatment</code>: data frame annotating all the compounds
investigated in the study.</li>
<li>
<code>tissue</code>: data.frame annotating all tissues investigated
in the study.</li>
</ul>
</div>
<div class="section level3">
<h3 id="molecularprofiles">
<code>molecularProfiles</code>:<a class="anchor" aria-label="anchor" href="#molecularprofiles"></a>
</h3>
<ul>
<li>List of <code>SummarizedExperiment</code> objects containing the
molecular profiles of the cell lines, such as mutations, gene
expressions, or copy number variations.</li>
</ul>
<p>The molecular profiles slot of a PharmacoSet object contains all the
molecular expression data profiled in the dataset. Each type of data is
kept in a separate SummarizedExperiment (SE) object, and they are all
stored as a list. However, to insure coordination between data types,
there are specific annotations that must be included in the
SummarizedExperiment object. The annotations also differ slightly in the
cases of <code>perturbation</code> and <code>sensitivity</code>
datasets.</p>
<p>First of all, each SummarizedExperiment object must be labelled with
the type of dataincluded within the object. Currently only
<code>mutation</code>, <code>fusion</code>, and <code>rna</code>
molecular data types are recognized and handled differently by
summarizing and signature generating functions. This labelling is done
by calling, for example:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SE</span><span class="op">@</span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="st">"rna"</span></span></code></pre></div>
<p>The phenoData in each SummarizedExperiment object also needs to
contain specific columns labelling each experiment. For sensitivity type
datasets, this has to include:</p>
<p><code>sampleid</code>: This column contains a cell identifier that
matches <strong>exactly</strong> the rownames of the <code>sample</code>
slot of the PSet. <code>batchid</code>: This column contains and id of
the batch of each experiment. If such data is not available then it is
customary to fill with NA, but this column must be present.</p>
<p>Additionally, for the perturbation type datasets, there are several
more columns of metadata that must be included to enable the modelling
of differential gene expression. These are:</p>
<p><code>treatmentid</code>: The identifier of the drug used in each
experiment when a compound was applied. For controls, it should be left
as NA. These must match <strong>exactly</strong> to the rownames of the
<code>treatment</code> slot in the PharmacoSet.</p>
<p><code>duration</code>: If available, the ength of the experiment.
Should be 0 for controls.</p>
<p><code>concentration</code>: For experiments where compounds were
applied, the concentration. Should be 0 for controls.</p>
<p><code>xptype</code>: A label of either <code>perturbation</code>, of
<code>control</code> respectively.</p>
<p>Otherwise each SummarizedExperiment object should be constructed as
specified by the SummarizedExperiment package.</p>
<p>Once these SummarizedExperiment objects are created, they should be
added to a list, with the name of each object in the list being
descriptive of the type of data contained.</p>
</div>
<div class="section level3">
<h3 id="datasettype">
<code>datasetType</code><a class="anchor" aria-label="anchor" href="#datasettype"></a>
</h3>
<p>This slot is being deprecated, since <code>perturbation</code> data
can effectively before stored in the <code>colData</code> slot of the
<code>SummarizedExperiment</code> in the <code>molecularProfiles</code>
slot</p>
<p>The <code>datasetType</code> slot is a character string which
signifies whether the PharmacoSet contains drug dose
<code>sensitivity</code> data, genomic <code>perturbation</code> data,
or <code>both</code>. This slot is necessary for the subsetting and
intersecting function to be aware of what data types they should be
looking for.</p>
</div>
<div class="section level3">
<h3 id="treatmentresponse">
<code>treatmentResponse</code>:<a class="anchor" aria-label="anchor" href="#treatmentresponse"></a>
</h3>
<p>Previously called <code>@sensitivity</code> this slot contains
viability measurements for cell-lines treated with different therapeutic
compounds.</p>
<ul>
<li>
<code>n</code>: Number of experiments for each cell line treated
with a given drug.</li>
<li>
<code>info</code>: Metadata for each pharmacological
experiment.</li>
<li>
<code>raw</code>: All cell viability measurements at each drug
concentration from the drug dose-response curves.</li>
<li>
<code>profiles</code>: Drug sensitivity values summarizing each
dose-response curve (IC<span class="math inline">\(_{50}\)</span>, AUC,
etc.)</li>
</ul>
<p>This slot is currently being redesigned to support drug combination
experiments. See the <code>LongTable</code> vignette for information
about the new data strucutre.</p>
<p>This is a list containing all the data pertaining to drug dose
response experiments, usually only present in <code>sensitivity</code>
or <code>both</code> type datasets. This includes the names the
following names slots in the list:</p>
<p><code>info</code>: Metadata for each pharmacological experiment
stored in a <code>data.frame</code>. Each row of this data.frame should
be labelled with a unique experiment identifier. This data.frame must
also contain specific columns:</p>
<p><code>sampleid</code>: Contains a cell identifier that matches
<strong>exactly</strong> the rownames of the cell slot of the PSet.</p>
<p><code>treatmentid</code>: The identifier of the drug used in each
experiment when a compound was applied. These must match
<strong>exactly</strong> to the rownames of the drug slot in the
PharmacoSet.</p>
<p><code>raw</code>: This is a 3-D <code>array</code> of raw drug dose
response data. The first dimension is the experiments, with the names of
each row in this dimension labelled exactly as the experiment ids used
in the metadata data.frame <code>info</code> above. The second dimension
is the doses, with each “column” labelled as <code>doses</code>, where
the number is the column number. This is as long as there are different
number of doses for each experiment. The third dimension is always fixed
at 2, with names <code>Dose</code>, and <code>Viability</code>.
<code>Dose</code> contains the actual dose used (usually in
micro-molar), while viability contains the cell viability measurement at
that dose. Therefore, for each experiment there are two vectors, one of
doses and one of viabilities at each dose.</p>
<p><code>profiles</code> This is a data.frame containing down the rows,
with <strong>exactly</strong> matching rownames, the experiments as
labelled in the <code>info</code> data.frame, and each column being a
summary of the drug dose sensitivity, such as
<code>auc\_published</code> or <code>ic50\_published</code>, as
published with the data.</p>
<p><code>n</code> This is a <code>matrix</code> containing cellids
matching exactly the rownames of the cell slot of the PSet down the
rows, and drugids matching exactly the rownames of the drug slot in the
columns. This data.frame summaries how many experiments are in the data
for each pair, allowing quick identification of what pairs were tested
together. Note that this does not need to be generated before
constructing the object, it is generated by the constructor.</p>
</div>
<div class="section level3">
<h3 id="perturbation">
<code>perturbation</code>:<a class="anchor" aria-label="anchor" href="#perturbation"></a>
</h3>
<p>This slot is being deprecated, as discussed above in
<code>datasetType</code>.</p>
<ul>
<li>
<code>n</code>: Number of experiments for each cell line perturbed
by a given drug, for each molecular data type</li>
<li>
<code>profiles</code>: The metadata for the perturbation experiments
is available for each molecular type by calling the appropriate info
function</li>
</ul>
<p><code>curation</code>: list of data frames containing the mapping
between original drug, cell line, and tissue names to standardized
identifiers.</p>
<p>This slot is fully filled by the constructor, so nothing needs to be
created for it.</p>
<p><code>n</code>: This is a 3-D <code>array</code> containing cellids
matching exactly the rownames of the cell slot of the PSet down the
rows, and drugids matching exactly the rownames of the drug slot in the
columns, and the third dimension being the names of the different
molecular profile types. This array summarizes how many perturbation
experiments are in each molecular data type for each pair, allowing
quick identification of what pairs were tested together.</p>
<p><code>info</code>: This is always exactly the string referring to
each data type separately: ‘The metadata for the perturbation
experiments is available for each molecular type by calling the
appropriate info function’</p>
</div>
<div class="section level3">
<h3 id="annotation">
<code>annotation</code><a class="anchor" aria-label="anchor" href="#annotation"></a>
</h3>
<p>The <code>annotation</code> slot contains information pertaining to
the <code>PharmacoSet</code> as a whole, including the full information
on what packages were loaded into R when the PharmacoSet was created, as
well as the date and call to the constructor. The only data that must be
entered upon creation of the object is the name of the PharmacoSet,
which is passed to the constructor.</p>
<ul>
<li>
<code>name</code>: Acronym of the pharmacogenomic dataset.</li>
<li>
<code>dateCreated</code>: When the object was created.</li>
<li>
<code>sessionInfo</code>: Software environment used to create the
object.</li>
<li>
<code>call</code>: Set of parameters used to create the object.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="creating-the-pharmacoset">Creating the PharmacoSet<a class="anchor" aria-label="anchor" href="#creating-the-pharmacoset"></a>
</h2>
<p>Once all the data is in the proper format, it can be passed to a the
constructor function as below:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PharmacoSet.html">PharmacoSet</a></span><span class="op">(</span><span class="va">name</span>,</span>
<span>            molecularProfiles<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>            sample<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>            treatment<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>            sensitivityInfo<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>            sensitivityRaw<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/array.html" class="external-link">array</a></span><span class="op">(</span>dim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span>,</span>
<span>            sensitivityProfiles<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>            curationTreatment<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>            curationSample<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>            curationTissue<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>            datasetType<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sensitivity"</span>, <span class="st">"perturbation"</span>, <span class="st">"both"</span><span class="op">)</span>,</span>
<span>            verify <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Here, <code>name</code> is the name of the PharmacoSet as described
in annotations above, <code>molecularProfiles</code> is the list of
ExpressionSet objects described above, cell and drug are the data.frames
in the cell and drug slots. The <code>sensitivityInfo</code>,
<code>sensitivityRaw</code>, <code>sensitivityProfiles</code> are the
elements of the <code>sensitivity</code> slot, and curationTreatment,
curationSample, and curationTissue the contents of the curation slot.
Finally, <code>datasetType</code> is a string to signify which type of
data isincluded in the datset, and verify is a flag for running the
verification of the pSet after its construction. Any data detailed above
but missing from this constructor call will be created by the
constructor, and the default empty values are as in the call above.
Notice that none of the values should be set as NA, they should be
omited if the data is not being provided.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Petr Smirnov, Christopher Eeles, Zhaleh Safikhani, Mark Freeman, Feifei Li, Benjamin Haibe-Kains.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
