<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="PharmacoGx">
<title>PharmacoGx: An R Package for Analysis of Large Pharmacogenomic Datasets • PharmacoGx</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.3/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.3/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="PharmacoGx: An R Package for Analysis of Large Pharmacogenomic Datasets">
<meta property="og:description" content="PharmacoGx">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">PharmacoGx</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.1.7</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/PharmacoGx.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/CreatingPharmacoSet.html">PharmacoGx: An R Package for Analysis of Large Pharmacogenomic Datasets</a>
    <a class="dropdown-item" href="../articles/DetectingDrugSynergyAndAntagonism.html">Detecting Drug Synergy and Antagonism with PharmacoGx 3.0+</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/bhklab/PharmacoGx/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>PharmacoGx: An R Package for Analysis of Large Pharmacogenomic Datasets</h1>
                        <h4 data-toc-skip class="author">Petr
Smirnov</h4>
            <address class="author_afil">
      Princess Margaret Cancer Centre, University Health Network,
Toronto Canada<br><a class="author_email" href="mailto:#"></a><a href="mailto:petr.smirnov@uhnresearch.ca" class="email">petr.smirnov@uhnresearch.ca</a>
      </address>
                              <h4 data-toc-skip class="author">Christopher
Eeles</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:christopher.eeles@uhnresearch.ca" class="email">christopher.eeles@uhnresearch.ca</a>
      
                              <h4 data-toc-skip class="author">Zhaleh
Safikhani</h4>
            <address class="author_afil">
      Princess Margaret Cancer Centre, University Health Network,
Toronto CanadaDepartment of Medical Biophysics, University of Toronto,
Toronto
Canada<br><h4 data-toc-skip class="author">Benjamin
Haibe-Kains</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:benjamin.haibe.kains@utoronto.ca" class="email">benjamin.haibe.kains@utoronto.ca</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/bhklab/PharmacoGx/blob/HEAD/vignettes/PharmacoGx.Rmd" class="external-link"><code>vignettes/PharmacoGx.Rmd</code></a></small>
      <div class="d-none name"><code>PharmacoGx.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Pharmacogenomics holds much potential to aid in discovering drug
response biomarkers and developing novel targeted therapies, leading to
development of precision medicine and working towards the goal of
personalized therapy. Several large experiments have been conducted,
both to molecularly characterize drug dose response across many cell
lines, and to examine the molecular response to drug administration.
However, the experiments lack a standardization of protocols and
annotations, hindering meta-analysis across several experiments.</p>
<p><code>PharmacoGx</code> was developed to address these challenges, by
providing a unified framework for downloading and analyzing large
pharmacogenomic datasets which are extensively curated to ensure maximum
overlap and consistency. <code>PharmacoGx</code> is based on a level of
abstraction from the raw experimental data, and allows bioinformaticians
and biologists to work with data at the level of genes, drugs and cell
lines. This provides a more intuitive interface and, in combination with
unified curation, simplifies analyses between multiple datasets.</p>
<p>To organize the data released by each experiment, we developed the
<em>PharmacoSet</em> class. This class efficiently stores different
types of data and facilitates interrogating the data by drug or cell
line. The <em>PharmacoSet</em> is also versatile in its ability to deal
with two distinct types of pharmacogenomic datasets. The first type,
known as <em>sensitivity</em> datasets, are datasets where cell lines
were profiled on the molecular level, and then tested for drug dose
response. The second type of dataset is the <em>perturbation</em>
dataset. These types of datasets profile a cell line on the molecular
level before and after administration of a compound, to characterize the
action of the compound on the molecular level.</p>
<p>With the first release of <em>PharmacoGx</em> we have fully curated
and created PharmacoSet objects for three publicly available large
pharmacogenomic datasets. Two of these datasets are of the
<em>sensitivity</em> type. These are the Genomics of Drug Sensitivity in
Cancer Project (GDSC) <span class="citation">(Garnett et al.
2012)</span> and the Cancer Cell Line Encyclopedia (CCLE) <span class="citation">(Barretina et al. 2012)</span>. The third dataset is of
the <em>perturbation</em> type, the Connectivity Map (CMAP) project
<span class="citation">(Lamb et al. 2006)</span>.</p>
<p>Furthermore, <code>PharmacoGx</code> provides a suite of parallelized
functions which facilitate drug response biomarker discovery, and
molecular drug characterization. This vignette will provide two example
analysis case studies. The first will be comparing gene expression and
drug sensitivity measures across the CCLE and GDSC projects. The second
case study will interrogate the CMAP database with a known signature of
up and down regulated genes for HDAC inhibitors as published in [<span class="citation">Glaser et al. (2003)</span>}. Using the Connectivity
Score as defined in <span class="citation">(Lamb et al. 2006)</span>, it
will be seen that known HDAC inhibitors have a high numerical score and
high significance.</p>
<p>For the purpose of this vignette, an extremely minuscule subset of
all three <em>PharmacoSet</em> objects are included with the package as
example data. They are included for illustrative purposes only, and the
results obtained with them will likely be meaningless.</p>
<div class="section level3">
<h3 id="installation-and-settings">Installation and Settings<a class="anchor" aria-label="anchor" href="#installation-and-settings"></a>
</h3>
<p><code>PharmacoGx</code> requires that several packages are installed.
However, all dependencies are available from CRAN or Bioconductor.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">'PharmacoGx'</span><span class="op">)</span></span></code></pre></div>
<p>Load <code>PharamacoGx</code> into your current workspace:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">PharmacoGx</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="requirements">Requirements<a class="anchor" aria-label="anchor" href="#requirements"></a>
</h3>
<p><code>PharmacoGx</code> has been tested on Windows, Mac and Cent OS
platforms. The packages uses the core R package <code>parallel</code> to
preform parallel computations, and therefore if parallelization is
desired, the dependencies for the parallel package must be met.</p>
</div>
</div>
<div class="section level2">
<h2 id="downloading-pharmacoset-objects">Downloading PharmacoSet Objects<a class="anchor" aria-label="anchor" href="#downloading-pharmacoset-objects"></a>
</h2>
<p>We have made the PharmacoSet objects of the curated datasets
available for download using functions provided in the package. A table
of available PharmacoSet objects can be obtained by using the
<code><a href="../reference/availablePSets.html">availablePSets()</a></code> function. Any of the PharmacoSets in the
table can then be downloaded by calling <em>downloadPSet</em>, which
saves the datasets into a directory of the users choice, and returns the
data into the R session.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/availablePSets.html">availablePSets</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">GDSC</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/downloadPSet.html">downloadPSet</a></span><span class="op">(</span><span class="st">"GDSC_2020(v2-8.2)"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="downlading-drug-signatures">Downlading Drug Signatures<a class="anchor" aria-label="anchor" href="#downlading-drug-signatures"></a>
</h3>
<p>The package also provides tools to compute drug perturbation and
sensitivity signatures, as explained below. However, the computation of
the perturbation signatures is very lengthy, so for users’ convenience
we have precomputed the signatures for our perturbation PharmacoSet
objects and made them available for download using the function
<code><a href="../reference/downloadPertSig.html">downloadPertSig()</a></code>.</p>
</div>
</div>
<div class="section level2">
<h2 id="case-study">Case Study<a class="anchor" aria-label="anchor" href="#case-study"></a>
</h2>
<div class="section level3">
<h3 id="inconsistency-across-large-pharmacogenomic-studies">(In)Consistency across large pharmacogenomic studies<a class="anchor" aria-label="anchor" href="#inconsistency-across-large-pharmacogenomic-studies"></a>
</h3>
<p>Our first case study illustrates the functions for analysis of the
<em>sensitivity</em> type of dataset. The case study will investigate
the consistency between the GDSC and CCLE datasets, recreating the
analysis similar to our <em>Inconsistency in Large Pharmacogenomic
Studies</em> paper <span class="citation">(Haibe-Kains et al.
2013)</span>. In both CCLE and GDSC, the transcriptome of cells was
profiled using an Affymatrix microarray chip. Cells were also tested for
their response to increasing concentrations of various compounds, and
form this the IC50 and AUC were computed. However, the cell and drugs
names used between the two datasets were not consistent. Furthermore,
two different microarray platforms were used. However,
<code>PharmacoGx</code> allows us to overcome these differences to do a
comparative study between these two datasets.</p>
<p>GDSC was profiled using the hgu133a platform, while CCLE was profiled
with the expanded hgu133plus2 platform. While in this case the hgu133a
is almost a strict subset of hgu133plus2 platform, the expression
information in <code>PharmacoSet</code> objects is summarized by
Ensemble Gene Ids, allowing datasets with different platforms to be
directly compared. The probe to gene mapping is done using the
BrainArray customCDF for each platform <span class="citation">(Sabatti,
Karsten, and Geschwind 2002)</span>.</p>
<p>To begin, you would load the datasets from disk or download them
using the <em>downloadPSet</em> function above. In the following
example, we use the toy datasets provided with the package to illustrate
the process, but to recreate the full analysis the full PharmacoSets
have to be downloaded.</p>
<p>We want to investigate the consistency of the data between the two
datasets. The common intersection between the datasets can then be found
using <em>intersectPSet</em>. We create a summary of the gene expression
and drug sensitivity measures for both datasets, so we are left with one
gene expression profile and one sensitivity profile per cell line within
each dataset. We can then compare the gene expression and sensitivity
measures between the datasets using a standard correlation
coefficient.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biobase" class="external-link">Biobase</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/S4Vectors" class="external-link">S4Vectors</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">PharmacoGx</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"GDSCsmall"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"CCLEsmall"</span><span class="op">)</span></span>
<span>  <span class="va">commonGenes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="../reference/PharmacoSet-accessors.html">fNames</a></span><span class="op">(</span><span class="va">GDSCsmall</span>, <span class="st">"rna"</span><span class="op">)</span>,</span>
<span>                           <span class="fu"><a href="../reference/PharmacoSet-accessors.html">fNames</a></span><span class="op">(</span><span class="va">CCLEsmall</span>,<span class="st">"rna"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">common</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/intersectPSet.html">intersectPSet</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'CCLE'</span><span class="op">=</span><span class="va">CCLEsmall</span>,</span>
<span>                               <span class="st">'GDSC'</span><span class="op">=</span><span class="va">GDSCsmall</span><span class="op">)</span>,</span>
<span>                          intersectOn<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cell.lines"</span>, <span class="st">"drugs"</span><span class="op">)</span>,</span>
<span>                          strictIntersect<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span></span>
<span>  <span class="va">GDSC.auc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CoreGx/man/summarizeSensitivityProfiles.html" class="external-link">summarizeSensitivityProfiles</a></span><span class="op">(</span></span>
<span>                <span class="va">common</span><span class="op">$</span><span class="va">GDSC</span>,</span>
<span>                sensitivity.measure<span class="op">=</span><span class="st">'auc_published'</span>,</span>
<span>                summary.stat<span class="op">=</span><span class="st">"median"</span>,</span>
<span>                verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="va">CCLE.auc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CoreGx/man/summarizeSensitivityProfiles.html" class="external-link">summarizeSensitivityProfiles</a></span><span class="op">(</span></span>
<span>                <span class="va">common</span><span class="op">$</span><span class="va">CCLE</span>,</span>
<span>                sensitivity.measure<span class="op">=</span><span class="st">'auc_published'</span>,</span>
<span>                summary.stat<span class="op">=</span><span class="st">"median"</span>,</span>
<span>                verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">GDSC.ic50</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CoreGx/man/summarizeSensitivityProfiles.html" class="external-link">summarizeSensitivityProfiles</a></span><span class="op">(</span></span>
<span>                <span class="va">common</span><span class="op">$</span><span class="va">GDSC</span>,</span>
<span>                sensitivity.measure<span class="op">=</span><span class="st">'ic50_published'</span>,</span>
<span>                summary.stat<span class="op">=</span><span class="st">"median"</span>,</span>
<span>                verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="va">CCLE.ic50</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CoreGx/man/summarizeSensitivityProfiles.html" class="external-link">summarizeSensitivityProfiles</a></span><span class="op">(</span></span>
<span>                <span class="va">common</span><span class="op">$</span><span class="va">CCLE</span>,</span>
<span>                sensitivity.measure<span class="op">=</span><span class="st">'ic50_published'</span>,</span>
<span>                summary.stat<span class="op">=</span><span class="st">"median"</span>,</span>
<span>                verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">GDSCexpression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CoreGx/man/summarizeMolecularProfiles.html" class="external-link">summarizeMolecularProfiles</a></span><span class="op">(</span><span class="va">common</span><span class="op">$</span><span class="va">GDSC</span>,</span>
<span>                                        <span class="fu"><a href="../reference/PharmacoSet-accessors.html">cellNames</a></span><span class="op">(</span><span class="va">common</span><span class="op">$</span><span class="va">GDSC</span><span class="op">)</span>,</span>
<span>                                        mDataType<span class="op">=</span><span class="st">"rna"</span>,</span>
<span>                                        features<span class="op">=</span><span class="va">commonGenes</span>,</span>
<span>                                        verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="va">CCLEexpression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CoreGx/man/summarizeMolecularProfiles.html" class="external-link">summarizeMolecularProfiles</a></span><span class="op">(</span><span class="va">common</span><span class="op">$</span><span class="va">CCLE</span>,</span>
<span>                                         <span class="fu"><a href="../reference/PharmacoSet-accessors.html">cellNames</a></span><span class="op">(</span><span class="va">common</span><span class="op">$</span><span class="va">CCLE</span><span class="op">)</span>,</span>
<span>                                         mDataType<span class="op">=</span><span class="st">"rna"</span>,</span>
<span>                                         features<span class="op">=</span><span class="va">commonGenes</span>,</span>
<span>                                         verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="va">gg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PharmacoSet-accessors.html">fNames</a></span><span class="op">(</span><span class="va">common</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="st">'rna'</span><span class="op">)</span></span>
<span>  <span class="va">cc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PharmacoSet-accessors.html">cellNames</a></span><span class="op">(</span><span class="va">common</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">ge.cor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">cc</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">x</span>, <span class="va">d1</span>, <span class="va">d2</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">d1</span><span class="op">[</span> , <span class="va">x</span><span class="op">]</span>, <span class="va">d2</span><span class="op">[</span> , <span class="va">x</span><span class="op">]</span>, method<span class="op">=</span><span class="st">"spearman"</span>,</span>
<span>                use<span class="op">=</span><span class="st">"pairwise.complete.obs"</span><span class="op">)</span></span>
<span>  <span class="co">## TO DO:: Ensure all assays are name so we can call by name instead of index</span></span>
<span>  <span class="op">}</span>, d1<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">GDSCexpression</span>, <span class="fl">1</span><span class="op">)</span>, d2<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">CCLEexpression</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">ic50.cor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">cc</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">x</span>, <span class="va">d1</span>, <span class="va">d2</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">d1</span><span class="op">[</span>, <span class="va">x</span><span class="op">]</span>, <span class="va">d2</span><span class="op">[</span> , <span class="va">x</span><span class="op">]</span>, method<span class="op">=</span><span class="st">"spearman"</span>,</span>
<span>                use<span class="op">=</span><span class="st">"pairwise.complete.obs"</span><span class="op">)</span></span>
<span>  <span class="op">}</span>, d1<span class="op">=</span><span class="va">GDSC.ic50</span>, d2<span class="op">=</span><span class="va">CCLE.ic50</span><span class="op">)</span></span>
<span>  <span class="va">auc.cor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">cc</span>, <span class="kw">function</span> <span class="op">(</span><span class="va">x</span>, <span class="va">d1</span>, <span class="va">d2</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">d1</span><span class="op">[</span> , <span class="va">x</span><span class="op">]</span>, <span class="va">d2</span><span class="op">[</span> , <span class="va">x</span><span class="op">]</span>, method<span class="op">=</span><span class="st">"spearman"</span>,</span>
<span>                use<span class="op">=</span><span class="st">"pairwise.complete.obs"</span><span class="op">)</span></span>
<span>  <span class="op">}</span>, d1<span class="op">=</span><span class="va">GDSC.auc</span>, d2<span class="op">=</span><span class="va">CCLE.auc</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">w1</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/wilcox.test.html" class="external-link">wilcox.test</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">ge.cor</span>, y<span class="op">=</span><span class="va">auc.cor</span>,</span>
<span>                           conf.int<span class="op">=</span><span class="cn">TRUE</span>, exact<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="va">w2</span> <span class="op">&lt;-</span> <span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/wilcox.test.html" class="external-link">wilcox.test</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">ge.cor</span>, y<span class="op">=</span><span class="va">ic50.cor</span>,</span>
<span>                           conf.int<span class="op">=</span><span class="cn">TRUE</span>, exact<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="va">yylim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="va">ss</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"GE vs. AUC = %.1E\nGE vs. IC50 = %.1E"</span>,</span>
<span>                <span class="va">w1</span><span class="op">$</span><span class="va">p.value</span>, <span class="va">w2</span><span class="op">$</span><span class="va">p.value</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"GE"</span><span class="op">=</span><span class="va">ge.cor</span>,</span>
<span>               <span class="st">"AUC"</span><span class="op">=</span><span class="va">auc.cor</span>,</span>
<span>               <span class="st">"IC50"</span><span class="op">=</span><span class="va">ic50.cor</span><span class="op">)</span>,</span>
<span>          main<span class="op">=</span><span class="st">"Concordance between cell lines"</span>,</span>
<span>          ylab<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">R</span><span class="op">[</span><span class="va">s</span><span class="op">]</span><span class="op">)</span>,</span>
<span>          sub<span class="op">=</span><span class="va">ss</span>,</span>
<span>          ylim<span class="op">=</span><span class="va">yylim</span>,</span>
<span>          col<span class="op">=</span><span class="st">"lightgrey"</span>,</span>
<span>          pch<span class="op">=</span><span class="fl">20</span>,</span>
<span>          border<span class="op">=</span><span class="st">"black"</span><span class="op">)</span></span></code></pre></div>
<p><img src="PharmacoGx_files/figure-html/fig1-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="query-the-connectivity-map">Query the Connectivity Map<a class="anchor" aria-label="anchor" href="#query-the-connectivity-map"></a>
</h3>
<p>The second case study illustrates the analysis of a perturbation type
datasets, where the changes in cellular molecular profiles are compared
before and after administering a compound to the cell line. Of these
datasets, we have currently curated and made available for download the
Connectivity Map (CMAP) dataset <span class="citation">(Lamb et al.
2006)</span>.</p>
<p>For this case study, we will recreate an analysis from the paper by
Lamb et al., in which a known signature for HDAC inhibitors <span class="citation">(Glaser et al. 2003)</span> is used to recover drugs in
the CMAP dataset that are also known HDAC inhibitors. For this example,
the package includes this signature, already mapped to the gene level,
and it can be loaded by calling <code>data(HDAC\_genes)</code>.</p>
<p>Once again, we load the dataset, downloading it if needed using
<code>downloadPSet</code> . We then recreate drug signatures for each
drug using the function <code>drugPerturbationSig</code> to preform
statistical modelling of the transcriptomic response to the application
of each drug. We then compare the observed up-regulated and
down-regulated genes to a the known HDAC signature, using the GSEA
connectivity score to determine the correlation between the two
signatures.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">PharmacoGx</span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rapporter.github.io/pander/" class="external-link">pander</a></span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">CMAPsmall</span><span class="op">)</span></span>
<span>  <span class="va">drug.perturbation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/drugPerturbationSig.html">drugPerturbationSig</a></span><span class="op">(</span><span class="va">CMAPsmall</span>,</span>
<span>                                           mDataType<span class="op">=</span><span class="st">"rna"</span>,</span>
<span>                                           verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">HDAC_genes</span><span class="op">)</span></span>
<span></span>
<span>  <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">drug.perturbation</span><span class="op">[</span>,,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"tstat"</span>, <span class="st">"fdr"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>               <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">HDAC</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="../reference/connectivityScore.html">connectivityScore</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>,</span>
<span>                                 y<span class="op">=</span><span class="va">HDAC</span><span class="op">[</span>,<span class="fl">2</span>,drop<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span>,</span>
<span>                               method<span class="op">=</span><span class="st">"fgsea"</span>, nperm<span class="op">=</span><span class="fl">100</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span>, HDAC<span class="op">=</span><span class="va">HDAC_genes</span><span class="op">)</span></span>
<span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Connectivity"</span>, <span class="st">"P Value"</span><span class="op">)</span></span>
<span>  <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span>  <span class="va">res</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">res</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span></span>
<span>  <span class="fu">pander</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pander/man/pandoc.table.return.html" class="external-link">pandoc.table</a></span><span class="op">(</span><span class="va">res</span>,</span>
<span>    caption<span class="op">=</span><span class="st">'Connectivity Score results for HDAC inhibitor gene signature.'</span>,</span>
<span>    style <span class="op">=</span> <span class="st">'rmarkdown'</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<caption>Connectivity Score results for HDAC inhibitor gene signature.
</caption>
<thead><tr class="header">
<th align="center"> </th>
<th align="center">Connectivity</th>
<th align="center">P Value</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><strong>vorinostat</strong></td>
<td align="center">0.9449</td>
<td align="center">0.003946</td>
</tr>
<tr class="even">
<td align="center"><strong>alvespimycin</strong></td>
<td align="center">0.8831</td>
<td align="center">0.009112</td>
</tr>
<tr class="odd">
<td align="center"><strong>acetylsalicylic acid</strong></td>
<td align="center">0</td>
<td align="center">1</td>
</tr>
<tr class="even">
<td align="center"><strong>rosiglitazone</strong></td>
<td align="center">-0.714</td>
<td align="center">0.01476</td>
</tr>
<tr class="odd">
<td align="center"><strong>pioglitazone</strong></td>
<td align="center">-0.7335</td>
<td align="center">0.1059</td>
</tr>
</tbody>
</table>
<p>As we can see, the known HDAC inhibitor Varinostat has a very strong
connectivity score, as well as a very high significance as determined by
permutation testing, in comparison to the other drugs, which score
poorly.</p>
<p>This example serves as a validation of the method, and demonstrates
the ease with which drug perturbation analysis can be done using
<code>PharmacoGx</code>. While in this case we were matching a drug
signature with a drug class signature, this method can also be used in
the discovery of drugs that are anti-correlated with known disease
signatures, to look for potential new treatments and drug
repurposing.</p>
</div>
</div>
<div class="section level2">
<h2 id="estimating-drug-sensitivity-measures">Estimating Drug Sensitivity Measures<a class="anchor" aria-label="anchor" href="#estimating-drug-sensitivity-measures"></a>
</h2>
<p><em>PharmacoGx</em> includes functions to calculate estimated AUC
(Area Under drug response Curve) and IC50 values from drugs dose
response experiments that measure cell viability after applications of
varying concentrations of drug. Additionally, these measures are
recomputed for every <em>sensitivity</em> <code>PharmacoSet</code> we
create and included alongside any measures published with the original
data. The AUC measures originally published are labelled as
<em>auc_published</em>, while the recomputed measures are labelled as
<em>auc_recomputed</em>, and likewise for the IC50.</p>
<p>While the <em>PharmacoSets</em> already contain the recomputed data,
the AUC and IC50 can be calculated for arbitrary data using the
<em>computeIC50</em> and <em>computeAUC</em> functions. The AUC can be
calculated using either the area under the curve defined by the actual
points recorded, or the area under the curve fitted to the data.</p>
<div class="section level3">
<h3 id="curve-fitting">Curve Fitting<a class="anchor" aria-label="anchor" href="#curve-fitting"></a>
</h3>
<p>While AUC can be numerically calculated without curve fitting, to
estimate the IC50 a drug dose response curve must be fit to the data.The
dose-response curves are fitted to the equation <span class="math display">\[
y = E_{\infty} +  \frac{ 1 - E_{\infty} }{1+(\frac{x}{IC50})^{HS}}
\]</span> where the maximum viability is normalized to be <span class="math inline">\(y = y(0) = 1\)</span>, <span class="math inline">\(E_{\infty}\)</span> denotes the minimum possible
viability achieved by administering any amount of the drug, <span class="math inline">\(IC50\)</span> is the concentration at which
viability is reduced to half of the viability observed in the presence
of an arbitrarily large concentration of drug, and <span class="math inline">\(HS\)</span> is a parameter describing the
cooperativity of binding. <span class="math inline">\(HS &lt; 1\)</span>
denotes negative binding cooperativity, <span class="math inline">\(HS =
1\)</span> denotes noncooperative binding, and <span class="math inline">\(HS &gt; 1\)</span> denotes positive binding
cooperativity. The parameters of the curves are fitted using the least
squares optimization framework. The fitting of these curves to arbitrary
points is implemented by the <em>logLogisticRegression</em>
function.</p>
</div>
<div class="section level3">
<h3 id="plotting-drug-dose-response-data">Plotting Drug-Dose Response Data<a class="anchor" aria-label="anchor" href="#plotting-drug-dose-response-data"></a>
</h3>
<p>Drug-Dose response data included in the PharmacoSet objects can be
conviniently plotted using the <code>drugDoseResponseCurve</code>
function. Given a list of PharmacoSets, a drug name and a cell name, it
will plot the drug dose response curves for the given cell-drug
combination in each dataset, allowing direct comparisons of data between
datasets.</p>
</div>
</div>
<div class="section level2">
<h2 id="gene-drug-association-modelling">Gene-Drug Association Modelling<a class="anchor" aria-label="anchor" href="#gene-drug-association-modelling"></a>
</h2>
<p><code>PharmacoGx</code> provides methods to model the association
between drugs and molecular data such as transcriptomics, genomics and
proteomics.</p>
<p><em>Sensitivity</em> studies allow the discovery of molecular
features that improve or inhibit the sensitivity of cell lines to
various compounds, by looking at the association between the expression
of the feature and the response towards each compound. This allows the
selection of drugs to be tailored to each specific patient based on the
expressed known sensitivity biomarkers. The function
<em>drugSensitivitySig</em> models these associations.</p>
<p><em>Perturbation</em> studies on the other hand look at the molecular
profiles of a cell before and after application of a drug, and therefore
can characterize the action of a drug on the molecular level. It is
hypothesized that drugs which act to down-regulate expression of known
disease biomarkers would be effective in reversing the cell from a
diseased to healthy state. The function <em>drugPerturbationSig</em>
models the molecular profiles of drugs tested in a <em>perturbation</em>
dataset.</p>
<div class="section level3">
<h3 id="sensitivity-modelling">Sensitivity Modelling<a class="anchor" aria-label="anchor" href="#sensitivity-modelling"></a>
</h3>
<p>The association between molecular features and response to a given
drug is modelled using a linear regression model adjusted for tissue
source:</p>
<p><span class="math display">\[
Y = \beta_{0} + \beta_{i}G_i + \beta_{t}T + \beta_{b}B
\]</span></p>
<p>where <span class="math inline">\(Y\)</span> denotes the drug
sensitivity variable, <span class="math inline">\(G_i\)</span>, <span class="math inline">\(T\)</span> and <span class="math inline">\(B\)</span> denote the expression of gene i, the
tissue source and the experimental batch respectively, and <span class="math inline">\(\beta\)</span>s are the regression coefficients.
The strength of gene-drug association is quantified by <span class="math inline">\(\beta_i\)</span>, above and beyond the
relationship between drug sensitivity and tissue source. The variables Y
and G are scaled (standard deviation equals to 1) to estimate
standardized coefficients from the linear model. Significance of the
gene-drug association is estimated by the statistical significance of
<span class="math inline">\(\beta_i\)</span> (two-sided t test).
P-values are then corrected for multiple testing using the false
discovery rate (FDR) approach.</p>
<p>As an example of the efficacy of the modelling approach, we can model
the significance of the association between two drugs and their known
biomarkers in CCLE. We examine the association between drug
<em>17-AAG</em> and gene <em>NQO1</em>, as well as drug
<em>PD-0325901</em> and gene <em>BRAF</em>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rapporter.github.io/pander/" class="external-link">pander</a></span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">CCLEsmall</span><span class="op">)</span></span>
<span>  <span class="va">features</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PharmacoSet-accessors.html">fNames</a></span><span class="op">(</span><span class="va">CCLEsmall</span>, <span class="st">"rna"</span><span class="op">)</span><span class="op">[</span></span>
<span>                          <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="../reference/PharmacoSet-accessors.html">featureInfo</a></span><span class="op">(</span><span class="va">CCLEsmall</span>,</span>
<span>                                            <span class="st">"rna"</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol</span> <span class="op">==</span> <span class="st">"NQO1"</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">sig.rna</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CoreGx/man/drugSensitivitySig.html" class="external-link">drugSensitivitySig</a></span><span class="op">(</span>object<span class="op">=</span><span class="va">CCLEsmall</span>,</span>
<span>                            mDataType<span class="op">=</span><span class="st">"rna"</span>,</span>
<span>                            drugs<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"17-AAG"</span><span class="op">)</span>,</span>
<span>                            features<span class="op">=</span><span class="va">features</span>,</span>
<span>                            sensitivity.measure<span class="op">=</span><span class="st">"auc_published"</span>,</span>
<span>                            molecular.summary.stat<span class="op">=</span><span class="st">"median"</span>,</span>
<span>                            sensitivity.summary.stat<span class="op">=</span><span class="st">"median"</span>,</span>
<span>                            verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="va">sig.mut</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CoreGx/man/drugSensitivitySig.html" class="external-link">drugSensitivitySig</a></span><span class="op">(</span>object<span class="op">=</span><span class="va">CCLEsmall</span>,</span>
<span>                            mDataType<span class="op">=</span><span class="st">"mutation"</span>,</span>
<span>                            drugs<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PD-0325901"</span><span class="op">)</span>,</span>
<span>                            features<span class="op">=</span><span class="st">"BRAF"</span>,</span>
<span>                            sensitivity.measure<span class="op">=</span><span class="st">"auc_published"</span>,</span>
<span>                            molecular.summary.stat<span class="op">=</span><span class="st">"and"</span>,</span>
<span>                            sensitivity.summary.stat<span class="op">=</span><span class="st">"median"</span>,</span>
<span>                            verbose<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span>  <span class="va">sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">sig.rna</span>, <span class="va">sig.mut</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sig</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"17-AAG + NQO1"</span>,<span class="st">"PD-0325901 + BRAF"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sig</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/dimnames.html" class="external-link">dimnames</a></span><span class="op">(</span><span class="va">sig.mut</span><span class="op">)</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></span>
<span>  <span class="fu">pander</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pander/man/pandoc.table.return.html" class="external-link">pandoc.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">sig</span><span class="op">)</span>, style <span class="op">=</span> <span class="st">"rmarkdown"</span>, caption<span class="op">=</span><span class="st">'P Value of Gene-Drug Association'</span> <span class="op">)</span></span></code></pre></div>
<table class="table">
<caption>P Value of Gene-Drug Association</caption>
<thead><tr class="header">
<th align="center"> </th>
<th align="center">17-AAG + NQO1</th>
<th align="center">PD-0325901 + BRAF</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><strong>estimate</strong></td>
<td align="center">0.601</td>
<td align="center">0.8252</td>
</tr>
<tr class="even">
<td align="center"><strong>se</strong></td>
<td align="center">0.05368</td>
<td align="center">0.1339</td>
</tr>
<tr class="odd">
<td align="center"><strong>n</strong></td>
<td align="center">492</td>
<td align="center">472</td>
</tr>
<tr class="even">
<td align="center"><strong>tstat</strong></td>
<td align="center">11.2</td>
<td align="center">6.161</td>
</tr>
<tr class="odd">
<td align="center"><strong>fstat</strong></td>
<td align="center">125.3</td>
<td align="center">37.95</td>
</tr>
<tr class="even">
<td align="center"><strong>pvalue</strong></td>
<td align="center">6.034e-26</td>
<td align="center">1.613e-09</td>
</tr>
<tr class="odd">
<td align="center"><strong>df</strong></td>
<td align="center">469</td>
<td align="center">449</td>
</tr>
<tr class="even">
<td align="center"><strong>fdr</strong></td>
<td align="center">6.034e-26</td>
<td align="center">1.613e-09</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="perturbation-modelling">Perturbation Modelling<a class="anchor" aria-label="anchor" href="#perturbation-modelling"></a>
</h3>
<p>The molecular response profile of a given drug is modelled as a
linear regression model adjusted experimental batch effects, cell
specific differences, and duration of experiment to isolate the effect
of the concentration of the drug applied:</p>
<p><span class="math display">\[
G = \beta_{0} + \beta_{i}C_i + \beta_{t}T + \beta_{d}D + \beta_{b}B
\]</span></p>
<p>where <span class="math inline">\(G\)</span> denotes the molecular
feature expression (gene), <span class="math inline">\(C_i\)</span>
denotes the concentration of the compound applied, <span class="math inline">\(T\)</span> the cell line identity, <span class="math inline">\(D\)</span> denotes the duration of the experiment,
<span class="math inline">\(B\)</span> denotes the experimental batch,
and <span class="math inline">\(\beta\)</span>s are the regression
coefficients. The strength of feature response is quantified by <span class="math inline">\(\beta_i\)</span>. Unlike the sensitivity
signatures, the <span class="math inline">\(G\)</span> and <span class="math inline">\(C\)</span> variables are unscaled. Significance of
the gene-drug association is estimated by the statistical significance
of <span class="math inline">\(\beta_i\)</span>, calculated using an
F-test on the improvement in fit after the inclusion of the term.
P-values are then corrected for multiple testing using the false
discovery rate (FDR) approach.</p>
</div>
</div>
<div class="section level2">
<h2 id="connectivity-scoring">Connectivity Scoring<a class="anchor" aria-label="anchor" href="#connectivity-scoring"></a>
</h2>
<p>The package also provides two methods for quantifying the similarity
between two molecular signatures of the form returned by
<em>drugPerturbationSig</em> and <em>drugSensitivitySig</em>, or a set
of up and down regulated genes as part of a disease signature. The two
methods are the <em>GSEA</em> method as introduced by Subramanian et at
<span class="citation">(Subramanian et al. 2005)</span>, and
<em>GWC</em>, a method based on a weighted Spearman correlation
coefficient. Both methods are implemented by the
<em>connectivityScore</em> function.</p>
<div class="section level3">
<h3 id="gsea">GSEA<a class="anchor" aria-label="anchor" href="#gsea"></a>
</h3>
<p>The <em>GSEA</em> method is implemented to compare a signature
returned by <em>drugPerturbationSig</em> with a known set of up and down
regulated genes in a disease state. For the disease signature, the
function expects a vector of features with a value, either binary (1,
-1) or continuous, where the sign signifies if the gene is up or down
regulated in the disease. The names of the vector are expected to be the
feature names, matching the feature names used in the drug signature.
The function then returns a GSEA score measuring the concordance of the
disease signature to the drug signature, as well as an estimated P-Value
for the significance of the connectivity determined by permutation
testing.</p>
</div>
<div class="section level3">
<h3 id="gwc">GWC<a class="anchor" aria-label="anchor" href="#gwc"></a>
</h3>
<p>The GWC method (Genome Wide Correlation) is implemented to compare
two signatures of the same length, such as two drug signatures returned
by <em>drugPerturbationSig</em>. The score is a Spearman correlation
weighted by the normalized negative logarithm significance of each
value. The normalization is done so that datasets of different size can
be compared without concern for lower p-values in one dataset due to
larger sample sizes.</p>
<p>More precisely, take <span class="math inline">\(X_i\)</span> and
<span class="math inline">\(Y_i\)</span> to be the ranks of the first
and second set of data respectively, and <span class="math inline">\(Px_i\)</span>, <span class="math inline">\(Py_i\)</span> to be the p-values of those
observations. The weight for each pair of observations is:</p>
<p><span class="math display">\[
Wx_i = \frac{-\log_{10}(Px_i)}{\sum_{i}-\log_{10}(Px_i)}
\]</span></p>
<p><span class="math display">\[
Wy_i = \frac{-\log_{10}(Py_i)}{\sum_{i}-\log_{10}(Py_i)}
\]</span></p>
<p><span class="math display">\[
W_i = Wx_i + Wy_i
\]</span></p>
<p>If we further define the weighted mean as follows: <span class="math display">\[
m(X;W) = \frac{\sum_i W_i X_i}{\sum_i{W_i}}
\]</span></p>
<p>Then the weighted correlation is given by:</p>
<p><span class="math display">\[
cor(X,Y,W) = \frac{\frac{\sum_i W_i (X_i - m(X;W))(Y_i -
m(Y,W))}{\sum_i W_i}}{\sqrt{(\frac{\sum_i W_i (X_i - m(X;W))^2}{\sum_i
W_i})
(\frac{\sum_i W_i (Y_i - m(Y;W))^2}{\sum_i W_i})}}
\]</span></p>
<p>This correlation therefore takes into account not only the ranking of
each feature in a signature, but also of the significance of each
rank.</p>
</div>
</div>
<div class="section level2">
<h2 id="acknowledgements">Acknowledgements<a class="anchor" aria-label="anchor" href="#acknowledgements"></a>
</h2>
<p>The authors of the package would like to thank the investigators of
the Genomics of Drug Sensitivity in Cancer Project, the Cancer Cell Line
Encyclopedia and the Connectivity Map Project who have made their
invaluable data available to the scientific community. We would also
like to thank Mark Freeman for contributing the code for MLE fitting
drug-dose response curves towards this package. We are indebted to Nehme
El-Hachem, Donald Wang and Adrian She for their contributions towards
the accurate curation of the datasets. Finally, it is important to
acknowledge all the members of the Benjamin Haibe-Kains lab for their
contribution towards testing and feedback during the design of the
package.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set eval = FALSE if you don't want this info (useful for reproducibility) to appear</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code>R version 4.2.1 (2022-06-23)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.1 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] pander_0.6.5                PharmacoGx_3.1.7           
 [3] CoreGx_2.2.0                SummarizedExperiment_1.28.0
 [5] GenomicRanges_1.50.1        GenomeInfoDb_1.34.4        
 [7] IRanges_2.32.0              S4Vectors_0.36.1           
 [9] MatrixGenerics_1.10.0       matrixStats_0.63.0         
[11] Biobase_2.58.0              BiocGenerics_0.44.0        
[13] knitcitations_1.0.12        knitr_1.41                 
[15] BiocStyle_2.26.0           

loaded via a namespace (and not attached):
  [1] fgsea_1.24.0                colorspace_2.0-3           
  [3] ellipsis_0.3.2              rprojroot_2.0.3            
  [5] lsa_0.73.3                  XVector_0.38.0             
  [7] fs_1.5.2                    SnowballC_0.7.0            
  [9] DT_0.26                     MultiAssayExperiment_1.24.0
 [11] fansi_1.0.3                 lubridate_1.9.0            
 [13] RefManageR_1.4.0            xml2_1.3.3                 
 [15] codetools_0.2-18            cachem_1.0.6               
 [17] jsonlite_1.8.3              magicaxis_2.2.14           
 [19] cluster_2.1.3               shinydashboard_0.7.2       
 [21] shiny_1.7.3                 mapproj_1.2.9              
 [23] BiocManager_1.30.19         compiler_4.2.1             
 [25] httr_1.4.4                  backports_1.4.1            
 [27] Matrix_1.4-1                fastmap_1.1.0              
 [29] bench_1.1.2                 limma_3.54.0               
 [31] cli_3.4.1                   later_1.3.0                
 [33] visNetwork_2.1.2            htmltools_0.5.3            
 [35] tools_4.2.1                 igraph_1.3.5               
 [37] gtable_0.3.1                glue_1.6.2                 
 [39] GenomeInfoDbData_1.2.9      reshape2_1.4.4             
 [41] RANN_2.6.1                  dplyr_1.0.10               
 [43] maps_3.4.1                  fastmatch_1.1-3            
 [45] Rcpp_1.0.9                  slam_0.1-50                
 [47] jquerylib_0.1.4             pkgdown_2.0.6              
 [49] vctrs_0.5.1                 BumpyMatrix_1.6.0          
 [51] xfun_0.35                   stringr_1.5.0              
 [53] timechange_0.1.1            mime_0.12                  
 [55] lifecycle_1.0.3             gtools_3.9.4               
 [57] MASS_7.3-57                 zlibbioc_1.44.0            
 [59] scales_1.2.1                ragg_1.2.4                 
 [61] promises_1.2.0.1            relations_0.6-12           
 [63] parallel_4.2.1              RColorBrewer_1.1-3         
 [65] sets_1.0-21                 yaml_2.3.6                 
 [67] memoise_2.0.1               downloader_0.4             
 [69] ggplot2_3.4.0               sass_0.4.4                 
 [71] stringi_1.7.8               highr_0.9                  
 [73] NISTunits_1.0.1             desc_1.4.2                 
 [75] plotrix_3.8-2               checkmate_2.1.0            
 [77] caTools_1.18.2              boot_1.3-28                
 [79] BiocParallel_1.32.4         bibtex_0.5.0               
 [81] rlang_1.0.6                 pkgconfig_2.0.3            
 [83] systemfonts_1.0.4           bitops_1.0-7               
 [85] pracma_2.4.2                evaluate_0.18              
 [87] lattice_0.20-45             purrr_0.3.5                
 [89] htmlwidgets_1.5.4           tidyselect_1.2.0           
 [91] cowplot_1.1.1               plyr_1.8.8                 
 [93] magrittr_2.0.3              bookdown_0.30              
 [95] R6_2.5.1                    gplots_3.1.3               
 [97] generics_0.1.3              sm_2.2-5.7.1               
 [99] DelayedArray_0.24.0         pillar_1.8.1               
[101] RCurl_1.98-1.9              tibble_3.1.8               
[103] crayon_1.5.2                utf8_1.2.2                 
[105] KernSmooth_2.23-20          rmarkdown_2.18             
[107] grid_4.2.1                  data.table_1.14.6          
[109] marray_1.76.0               piano_2.14.0               
[111] digest_0.6.30               xtable_1.8-4               
[113] httpuv_1.6.6                textshaping_0.3.6          
[115] celestial_1.4.6             coop_0.6-3                 
[117] munsell_0.5.0               bslib_0.4.1                
[119] shinyjs_2.1.0              </code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-barretina_cancer_2012" class="csl-entry">
Barretina, Jordi, Giordano Caponigro, Nicolas Stransky, Kavitha
Venkatesan, Adam A Margolin, Sungjoon Kim, Christopher J Wilson, et al.
2012. <span>“The <span>Cancer</span> <span>Cell</span> <span>Line</span>
<span>Encyclopedia</span> Enables Predictive Modelling of Anticancer
Drug Sensitivity.”</span> <em>Nature</em> 483 (7391): 603–7.
</div>
<div id="ref-garnett_systematic_2012" class="csl-entry">
Garnett, Mathew J, Elena J Edelman, Sonja J Heidorn, Chris D Greenman,
Anahita Dastur, King Wai Lau, Patricia Greninger, et al. 2012.
<span>“Systematic Identification of Genomic Markers of Drug Sensitivity
in Cancer Cells.”</span> <em>Nature</em> 483 (7391): 570–75.
</div>
<div id="ref-glaser_gene_2003" class="csl-entry">
Glaser, Keith B, Michael J Staver, Jeffrey F Waring, Joshua Stender,
Roger G Ulrich, and Steven K Davidsen. 2003. <span>“Gene Expression
Profiling of Multiple Histone Deacetylase (<span>HDAC</span>)
Inhibitors: Defining a Common Gene Set Produced by <span>HDAC</span>
Inhibition in <span>T</span>24 and <span>MDA</span> Carcinoma Cell
Lines.”</span> <em>Molecular Cancer Therapeutics</em> 2 (2): 151–63.
</div>
<div id="ref-haibe-kains_inconsistency_2013" class="csl-entry">
Haibe-Kains, Benjamin, Nehme El-Hachem, Nicolai Juul Birkbak, Andrew C
Jin, Andrew H Beck, Hugo J W L Aerts, and John Quackenbush. 2013.
<span>“Inconsistency in Large Pharmacogenomic Studies.”</span>
<em>Nature</em> 504 (7480): 389–93.
</div>
<div id="ref-lamb_connectivity_2006" class="csl-entry">
Lamb, Justin, Emily D. Crawford, David Peck, Joshua W. Modell, Irene C.
Blat, Matthew J. Wrobel, Jim Lerner, et al. 2006. <span>“The
<span>Connectivity</span> <span>Map</span>: <span>Using</span>
<span>Gene</span>-<span>Expression</span> <span>Signatures</span> to
<span>Connect</span> <span>Small</span> <span>Molecules</span>,
<span>Genes</span>, and <span>Disease</span>.”</span> <em>Science</em>
313 (5795): 1929–35. <a href="https://doi.org/10.1126/science.1132939" class="external-link">https://doi.org/10.1126/science.1132939</a>.
</div>
<div id="ref-sabatti_thresholding_2002" class="csl-entry">
Sabatti, Chiara, Stanislav L Karsten, and Daniel H Geschwind. 2002.
<span>“<a href="https://www.ncbi.nlm.nih.gov/pubmed/11867081" class="external-link">Thresholding Rules
for Recovering a Sparse Signal from Microarray Experiments</a>.”</span>
<em>Mathematical Biosciences</em> 176 (1): 17–34.
</div>
<div id="ref-subramanian_gene_2005" class="csl-entry">
Subramanian, Aravind, Pablo Tamayo, Vamsi K. Mootha, Sayan Mukherjee,
Benjamin L. Ebert, Michael A. Gillette, Amanda Paulovich, et al. 2005.
<span>“Gene Set Enrichment Analysis: <span>A</span> Knowledge-Based
Approach for Interpreting Genome-Wide Expression Profiles.”</span>
<em>Proceedings of the National Academy of Sciences</em> 102 (43):
15545–50. <a href="https://doi.org/10.1073/pnas.0506580102" class="external-link">https://doi.org/10.1073/pnas.0506580102</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Petr Smirnov, Christopher Eeles, Zhaleh Safikhani, Mark Freeman, Feifei Li, Benjamin Haibe-Kains.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
